### PR TITLE
Fix build of toy ppx

### PR DESCRIPTION
**Pull request by shindere**
This is an example of how things could be made to work

### CODE REVIEW

File: ppx/Makefile [hunk 1]
```diff
@@ -1,11 +1,23 @@
...
+
+include_directories := $(addprefix -I , $(directories))
+
+flags := -g -nostdlib $(include_directories) \
+  -strict-sequence -safe-string -strict-formats \
+  -w +a-4-9-40-41-42-44-45-48 -warn-error A
+
+ocamlc := $(ROOTDIR)/runtime/ocamlrun $(ROOTDIR)/ocamlc $(flags)
+
+libraries := ocamlbytecomp.cma  ocamlcommon.cma
```
Comment (Lereena) [inline ppx/Makefile]: Why do we need these libraries? Because of Asttypes and so on in ppx?

File: ppx/Makefile [hunk 2]
```diff
@@ -1,11 +1,23 @@
...
 
 .PHONY: build
-build:	
-	cd printexc_ppx; $(ocamlc) -c printexc_ppx.ml
\ No newline at end of file
+build: printexc_ppx
+
+printexc_ppx: printexc_ppx.ml
+	$(ocamlc) -o $@ $(libraries) printexc_ppx.ml
```
Comment (Lereena) [inline ppx/Makefile]: "$@" here will be replaced by "printexc_ppx" without ".ml"?

**Comment by shindere** [comment]
Oxana Kostikova (2019/06/06 13:07 -0700):
> Why do we need these libraries? Because of Asttypes and so on in ppx?

Yes, exactly.

If you are able to see which commands Dune uses to compile your toy ppx,
you should be able to see that it, took, links in these libraries. Are
you able to check this?

Also, when you do
```bash
make world
```

the last command you see is the one used to build your toy ppx.

You could change to the ppx directory and try running the command
without the libraries, jsut to see what happens.

**Comment by shindere** [comment]
Oxana Kostikova (2019/06/06 13:10 -0700):
> Lereena commented on this pull request.
> 
> 
> 
> > \ No newline at end of file
> +build: printexc_ppx
> +
> +printexc_ppx: printexc_ppx.ml
> +	$(ocamlc) -o $@ $(libraries) printexc_ppx.ml
> 
> "$@" here will be replaced by "printexc_ppx" without ".ml"?

Yes, and more generally by the name of the target, i.e. what appears
before the `:`.

Similarly, `$^` can be used to represent all the prerequisites, i.e. the
list of all the things that appear after the target, and `$<` is
replaced by the first prerequisite, so you could use `$<` in your
Makefile to represente `printexc.ml` in the command.

**Comment by Lereena** [comment]
It seems I understood what is happening here, except for one thing. If one writes a similar file from scratch, how to determine what should be entered in the `directories`?

**Comment by shindere** [comment]
Oxana Kostikova (2019/06/06 13:55 -0700):
> It seems I understood what is happening here, except for one thing. If
> one writes a similar file from scratch, how to determine what should
> be entered in the `directories`?

Well that should be easy. Start with no directories at all and try to
compile.
You will see error messages about missing modules.
Use the `find` command to locate the module in the compiler sources and
add them.

Try again to compile, until your directory list is complete.

Does this make sense?

For instance I had to do that because I took the directory list from
ocamltest but it was missing the `Printtyp` module and I thus had to add
the `typing` directory.

**Comment by Lereena** [comment]
How to use the find command? For example, now it's missing Ast_mapper, I try just `find Ast_mapper` from the root directory and get nothing

**Comment by shindere** [comment]
Oxana Kostikova (2019/06/06 14:15 -0700):
> How to use the find command? For example, now it's missing Ast_mapper,
> I try just `find Ast_mapper` from the root directory and get nothing

It's excelelnt you are experimenting! Let me encourage youu to continue
this way!

Remember that OCaml puts the first letter of module names in uppercse
even when the files themselves start with a lower-case letter, so what
you are probably looking for is ast_mapper, with a lower-case a at the
beginning.

Then from the root directory of the ocaml sources you can do

```bash
find . -name '*ast_mapper*'
```

Also you can try to be "clever" and remember that ast is for abstract
syntax tree and then it may give you a hint that it's in parsing, but
try the find command anyway.

You may also look at its man page! Woudl you for instance be able to
discover how the `-iname` option may help you?

Oh and also, I wanted to ask you: which text editor are you using and,
more globally, which operating system are you using?

**Comment by Lereena** [comment]
Thank you very much, now I understand how to use it. Search without regard to text case looks rather useful. 

I use Manjaro Linux and VSCode ("Code - OSS" build to be precise)

**Comment by shindere** [comment]
Oxana Kostikova (2019/06/07 02:16 -0700):
> Thank you very much, now I understand how to use it. Search without
> regard to text case looks rather useful.

Good! I see you have read the documentation somehow. :)

> I use Manjaro Linux and VSCode ("Code - OSS" build to be precise)

Okay. I don't know how to configure this editor, but you may want to
configure it so that it does not let yousave files with trailing
whitespaces at end of lines. If you hae a look ty the PR I did on your
repository, you will see that some changes I had to do were about that.

One thing you could do, in addition to properly configuring your editor,
would be to run the following command from the root directory of your
copy of the OCaml repository:

```bash
cp tools/pre-commit-githook .git/hooks/pre-commit
```

This will install a git pre-commit hook. It's a script that is ran every
time you try to commit something and it will verify that no line that
you commit ends with trailing whitespaces, because this is considered
bad practice. It will also verify that your files end with a `\n`
(newline) character, plus a few other things.

As usually, do not hesitate to ask questions if something remains
unclear!

**Pull request merged**